Add easy scabbard circuits

leebradley opened the pull request:
Create CircuitBuilder and ScabbardCircuitBuilder

Added the CircuitBuilder, which facilitates easy circuit creation.

Added the network.circuit_builder function to automatically create a
circuit builder with a reference to the current network and specific
nodes within that network.

Added ScabbardCircuitBuilder, which specifically implements service
creation for Scabbard.

---

Allow ability to pass payload builder

Modify the testing framework payload creation function to allow a
payload builder to be passed, and to avoid panics since it's going to be
indirectly exposed as part of our api.

---

Use circuit_builder in circuit lifecycle test

Updated circuit lifestyle test to use the new
network.circuit_builder(...) functionality.

Signed-off-by: Lee Bradley <bradley@bitwise.io>

shannynalayna commented on splinterd/tests/admin/payload.rs:
```diff
@@ -43,17 +46,60 @@ pub(in crate::admin) fn make_create_circuit_payload(
+        .expect("Unable to generate circuit request")
+}
+
+#[derive(Debug)]
```
Would suggest we split these into a separate `error` module in the admin crate

shannynalayna commented on splinterd/tests/framework/builders.rs:
```diff
@@ -0,0 +1,966 @@
+use std::error::Error;
+use std::fmt;
+
+/// `CircuitBuildError` represents any errors that may arise during the final circuit build step
```
Once again, this could become a full directory `builder/error.rs`, and then the rest of the builders defined within that directory. Seems like there are a lot of separate builders here

leebradley commented:
Ready for re-review. Major changes in response to feedback:
* Paired down errors to make better use of `InvalidArgument` and `InternalError`
* Split up `builder.rs` into its own module folder
* Converted `CircuitBuilder` from a trait to a struct
* Added `veil` and `unveil` methods
* Updated `CircuitBuilder::build` to return an object that contains service ids and Node references

vaporos commented on splinterd/tests/admin/payload.rs:
```diff
@@ -43,17 +43,39 @@ pub(in crate::admin) fn make_create_circuit_payload(
 ) -> Vec<u8> {
+    let circuit_request = setup_circuit(circuit_id, node_info, admin_keys);
+    complete_create_payload(requester, signer, circuit_request)
+        .expect("Unable to generate circuit request")
```
Consider propagating the error up to remove the expect() here.

vaporos commented on splinterd/tests/framework/builders/circuit.rs:
```diff
@@ -0,0 +1,657 @@
+        self
+    }
+
+    /// Build the circuit, consuming this `CircuitBuilder` and receiving a [`Cir`] back
```
Typo here 'Cir'

vaporos commented on splinterd/tests/framework/builders/mod.rs:
```diff
@@ -0,0 +1,40 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Framework of builders for integration testing
```
Consider renaming this module "circuit_builder". Although ScabbardCircuitBuilder isn't a CircuitBuilder, this still provides a better definition of the contents of the module.

More philosophically, is ScabbardCircuitBuilder really a ScabbardCircuitBuilderVeil? If so, this could make sense:

crate::circuit_bulder:: # pub use things from sub-modules
crate::circuit_buiider::builder # CircuitBuilder
crate::circuit_builder::error
crate::circuit_builder::veil:: # any shared veil traits, etc.
crate::circuit_builder::veil::scabbard # ScabbardCircuitBuilderVeil

vaporos commented on splinterd/tests/framework/builders/mod.rs:
```diff
@@ -0,0 +1,40 @@
+use splinterd::node::Node;
+
+/// A generic network
+pub trait Networkable {
```
Rename this something to do with getting/supplying nodes and it will make more sense--NodeCollection perhaps. There really isn't a compelling reason not to directly use Network--except to constrain the portion of the API being used by the circuit builder module (which is a good enough reason imo), so this should purely be named from circuit builder's pov.

leebradley commented:
Made requested changes:
* Propagated up error instead of using `expect` for `make_create_circuit_payload`
* Restructured crate to follow the following format:
```
crate::circuit_bulder:: # pub use things from sub-modules
crate::circuit_buiider::builder # CircuitBuilder
crate::circuit_builder::error
crate::circuit_builder::veil:: # any shared veil traits, etc.
crate::circuit_builder::veil::scabbard # ScabbardCircuitBuilderVeil
```
* Fixed comment typo
* Renamed `Networkable` trait to `NodeCollection`

vaporos commented on splinterd/tests/framework/circuit_builder/mod.rs:
```diff
@@ -0,0 +1,40 @@
+pub use self::builder::{CircuitBuilder, CircuitData, CircuitService};
+pub use self::error::{AddScabbardServiceError, CircuitBuildError};
+pub use self::veil::scabbard::ScabbardCircuitBuilderVeil;
+pub use splinter::admin::messages::CreateCircuit;
```
This should not be re-exported here.

leebradley replied:
That was a leftover from when CreateCircuit was the result of the build() function, but I changed the return type to a new struct after talking it over with Peter. So yeah, this is definitely not necessary.

vaporos commented on splinterd/tests/admin/circuit_disband.rs:
```diff
@@ -49,22 +50,33 @@ pub fn test_2_party_circuit_lifecycle() {
     let mut node_a_events = BlockingAdminServiceEventIterator::new(
         node_a
-            .admin_service_event_client(&format!("test_circuit_{}", &circuit_id))
+            .admin_service_event_client(&management_type)
```
Is this change intentional?

leebradley replied:
Yeah, I was just taking advantage of the fact that the management type for the circuit is returned by the circuit builder.

vaporos commented on splinterd/tests/admin/circuit_disband.rs:
```diff
@@ -20,6 +20,7 @@ use splinterd::node::{Node, RestApiVariant};
 
 use crate::admin::circuit_commit::{commit_2_party_circuit, commit_3_party_circuit};
 use crate::admin::payload::{make_circuit_disband_payload, make_circuit_proposal_vote_payload};
+use crate::framework::circuit_builder::{CircuitData, ScabbardCircuitBuilderVeil as Scabbard};
```
The rename here is unnecessarily confusing, suggest using ScabbardCircuitBuilderVeil directly.
